[SelectionDAG] Use `EVT::getIntegerVT` in `getBitcastedAnyExtOrTrunc`

shiltian opened the pull request:
`SelectionDAG::getBitcastedAnyExtOrTrunc` assumes that there is always a valid
integer type corresponding to another type, which is not always true when it
comes to vector type. For example, `<3 x i8>` doesn't have a corresponding
integer type.

Fix SWDEV-464698.

RKSimon commented on llvm/test/CodeGen/AMDGPU/no-corresponding-integer-type.ll:
```diff
@@ -0,0 +1,13 @@
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx942 %s -o -
+
+define void @no_corresponding_integer_type(i8 %arg, ptr addrspace(1) %ptr) {
```
I realise this is a crash test, but its still nice to use FileCheck and check the codegen

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -1468,13 +1468,31 @@ SDValue SelectionDAG::getZExtOrTrunc(SDValue Op, const SDLoc &DL, EVT VT) {
+    assert(NumElements < ExtNumElements);
+    MVT ExtType = MVT::getVectorVT(ElementType.getSimpleVT(), ExtNumElements);
+    SmallVector<SDValue, 4> Ops(ExtNumElements, getUNDEF(ElementType));
+    SDValue ExtVec = getNode(ISD::BUILD_VECTOR, DL, ExtType, Ops);
```
Why not getUNDEF(ExtType)? If you do need to handle implicit truncation inside the BUILD_VECTOR you can use getSplat / getSplatBuildVector instead.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -1468,13 +1468,30 @@ SDValue SelectionDAG::getZExtOrTrunc(SDValue Op, const SDLoc &DL, EVT VT) {
 
 SDValue SelectionDAG::getBitcastedAnyExtOrTrunc(SDValue Op, const SDLoc &DL,
-                                                 EVT VT) {
+                                                EVT VT) {
```
From the name of the function I'm not sure this should be doing all this fancy vector stuff. Should the caller be fixed instead?

shiltian replied:
What this function does is two-folded. Based on existing code, it converts `Op` to `VT` where the destination type can not be a vector. It first casts `Op` to its corresponding integer type, no matter what type `Op` is. Next, depending on the size difference between `Op` and `VT`, it either extends or truncates the integer. With that being said, IMO it is this function's responsibility to make sure the validity of each step. However, the problem is in the first step. We can't assume there is always a corresponding integer type. `Op` can be a type with regular size (no matter it is vector or not), but it can also be some type with weird size, like `v3i8`, or `v2f12`.

shiltian commented:
gentle ping +1

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -1468,13 +1468,30 @@ SDValue SelectionDAG::getZExtOrTrunc(SDValue Op, const SDLoc &DL, EVT VT) {
+    return getBitcast(VT, DestOp);
+  }
+
   DestOp = getBitcast(MVT::getIntegerVT(Size), Op);
```
Why are we using MVT here and not EVT?

shiltian replied:
That is really a wonderful point. `EVT::getIntegerVT` can actually avoid the crash! Thanks!

shiltian committed:
[SelectionDAG] Fix a false assumption that there will always be a valid integer type corresponding to a vector type

`SelectionDAG::getBitcastedAnyExtOrTrunc` assumes that there is always a valid
integer type corresponding to another type, which is not always true when it
comes to vector type. For example, `<3 x i8>` doesn't have a corresponding
integer type.

Fix SWDEV-464698.

RKSimon reviewed (approved):
LGTM
